engines/mtropolis/runtime.cpp

```diff
@@ -0,0 +1,7382 @@
+	_value.asEvent = value;
+}
+
+void DynamicValue::setVarReference(const VarReference &value) {
+	if (_type != DynamicValueTypes::kVariableReference)
+		clear();
+	_type = DynamicValueTypes::kVariableReference;
+	_value.asVarReference.guid = value.guid;
+	_value.asVarReference.source = &_str;
+	_str = *value.source;
+}
+
+void DynamicValue::setString(const Common::String &value) {
+	if (_type != DynamicValueTypes::kString)
+		clear();
+	_type = DynamicValueTypes::kString;
+	_str = value;
+}
+
+void DynamicValue::setBool(bool value) {
+	if (_type != DynamicValueTypes::kBoolean)
+		clear();
+	_type = DynamicValueTypes::kBoolean;
+	_value.asBool = value;
+}
+
+void DynamicValue::setList(const Common::SharedPtr<DynamicList> &value) {
+	if (_type != DynamicValueTypes::kList)
+		clear();
+	_type = DynamicValueTypes::kList;
+	_list = value;
+}
+
+void DynamicValue::setReadProxy(const DynamicValueReadProxy &readProxy) {
+	Common::SharedPtr<DynamicList> listRef = readProxy.containerList;	// Back up list ref in case this is a self-assign
+	if (_type != DynamicValueTypes::kReadProxy)
+		clear();
+	_type = DynamicValueTypes::kReadProxy;
+	_value.asReadProxy = readProxy.pod;
+	_list = listRef;
+}
+
+void DynamicValue::setWriteProxy(const DynamicValueWriteProxy &writeProxy) {
+	Common::SharedPtr<DynamicList> listRef = writeProxy.containerList; // Back up list ref in case this is a self-assign
+	if (_type != DynamicValueTypes::kWriteProxy)
+		clear();
+	_type = DynamicValueTypes::kWriteProxy;
+	_value.asWriteProxy = writeProxy.pod;
+	_list = listRef;
+}
+
+bool DynamicValue::roundToInt(int32 &outInt) const {
+	if (_type == DynamicValueTypes::kInteger) {
+		outInt = _value.asInt;
+		return true;
+	} else if (_type == DynamicValueTypes::kFloat) {
+		outInt = static_cast<int32>(floor(_value.asFloat + 0.5));
+		return true;
+	}
+
+	return false;
+}
+
+bool DynamicValue::convertToType(DynamicValueTypes::DynamicValueType targetType, DynamicValue &result) const {
+	if (_type == targetType) {
+		result = *this;
+		return true;
+	}
+
+	switch (_type) {
+	case DynamicValueTypes::kInteger:
+		return convertIntToType(targetType, result);
+	case DynamicValueTypes::kFloat:
+		return convertFloatToType(targetType, result);
+	case DynamicValueTypes::kBoolean:
+		return convertBoolToType(targetType, result);
+	default:
+		warning("Couldn't convert dynamic value from source type");
+		return false;
+	}
+}
+
+void DynamicValue::setObject(const ObjectReference &value) {
+	if (_type != DynamicValueTypes::kObject)
+		clear();
+	_type = DynamicValueTypes::kObject;
+	_obj = value;
+}
+
+void DynamicValue::setObject(const Common::WeakPtr<RuntimeObject> &value) {
+	setObject(ObjectReference(value));
+}
+
+void DynamicValue::swap(DynamicValue &other) {
+	internalSwap(_type, other._type);
+	internalSwap(_str, other._str);
+	internalSwap(_list, other._list);
+	internalSwap(_obj, other._obj);
+
+	ValueUnion tempValue;
+	memcpy(&tempValue, &_value, sizeof(ValueUnion));
```

Comment:
It shouldn't matter because presumably "other" and "this" are presumably properly aligned anyway, and tempValue doesn't even need to be aligned because it's just used as a memcpy source/dest.